aws/request/waiter.go

```diff
@@ -0,0 +1,248 @@
+)
+
+// WaitWithContext will make requests for the API operation using NewRequest to
+// build API requests. The request's response will be compared against the
+// Waiter's Acceptors to determine the successful state of the resource the
+// waiter is inspecting.
+//
+// The waiter will continue until the target state defined by the Acceptors,
+// or the max attempts expires.
+//
+// Will return the WaiterResourceNotReadyErrorCode error code if the waiter's
+// retryer ShouldRetry returns false. This normally will happen when the max
+// wait attempts expires.
+func (w *Waiter) WaitWithContext(ctx aws.Context) error {
+	for shouldRetry := true; shouldRetry; shouldRetry = w.Retryer.ShouldRetry() {
+		req, err := w.NewRequest(w.RequestOptions)
+		if err != nil {
+			waiterLogf(w.Logger, "unable to create request %v", err)
+			return err
+		}
+		req.Handlers.Build.PushBack(MakeAddToUserAgentFreeFormHandler("Waiter"))
+
+		err = req.Send()
+		for _, a := range w.Acceptors {
+			result := false
+			var vals []interface{}
+			switch a.Matcher {
+			case PathAllWaiterMatch, PathWaiterMatch:
+				// Require all matches to be equal for result to match
+				vals, _ = awsutil.ValuesAtPath(req.Data, a.Argument)
+				if len(vals) == 0 {
+					break
+				}
+				result = true
+				for _, val := range vals {
+					if !awsutil.DeepEqual(val, a.Expected) {
+						result = false
+						break
+					}
+				}
+			case PathAnyWaiterMatch:
+				// Only a single match needs to equal for the result to match
+				vals, _ = awsutil.ValuesAtPath(req.Data, a.Argument)
+				for _, val := range vals {
+					if awsutil.DeepEqual(val, a.Expected) {
+						result = true
+						break
+					}
+				}
+			case PathListWaiterMatch:
+				// ignored matcher
+			case StatusWaiterMatch:
+				s := a.Expected.(int)
+				result = s == req.HTTPResponse.StatusCode
+			case ErrorWaiterMatch:
+				if aerr, ok := err.(awserr.Error); ok {
+					result = aerr.Code() == a.Expected.(string)
+				}
+			default:
+				waiterLogf(w.Logger, "WARNING: Waiter %s encountered unexpected matcher: %s",
+					w.Name, a.Matcher)
+			}
+
+			if !result {
+				// If there was no matching result found there is nothing more to do
+				// for this response, retry the request.
+				continue
+			}
+
+			switch a.State {
+			case SuccessWaiterState:
+				// waiter completed
+				return nil
+			case FailureWaiterState:
+				// Waiter failure state triggered
+				return awserr.New("ResourceNotReady",
+					"failed waiting for successful resource state", err)
+			case RetryWaiterState:
+				// clear the error and retry the operation
+				err = nil
+			default:
+				waiterLogf(w.Logger, "WARNING: Waiter %s encountered unexpected state: %s",
+					w.Name, a.State)
+			}
+		}
+		if err != nil {
+			// Error occurred during current waiter call
+			return err
+		}
+
+		// Delay to wait before inspecting the resource again
+		delay := w.Retryer.RetryDelay()
+		if ctx != nil {
```

Comment:
Thanks for pointing this out. I think that makes sense. I'll update these new methods to require a context.